bin/install_uninstall.ml

```diff
@@ -269,69 +269,32 @@ module File_ops_real (W : sig
       Some { need_version = true; callback })
   ;;
 
-  let replace_sites ~(get_location : Section.t -> Package.Name.t -> Stdune.Path.t) dp =
-    match
-      List.find_map dp ~f:(function
-        | Dune_lang.List [ Atom (A "name"); Atom (A name) ] -> Some name
-        | _ -> None)
-    with
-    | None -> dp
-    | Some name ->
-      List.map dp ~f:(function
-        | Dune_lang.List ((Atom (A "sections") as sexp_sections) :: sections) ->
-          let sections =
-            List.map sections ~f:(function
-              | Dune_lang.List [ (Atom (A section) as section_sexp); _ ] ->
-                let path =
-                  get_location
-                    (Option.value_exn (Section.of_string section))
-                    (Package.Name.of_string name)
-                in
-                let open Dune_lang.Encoder in
-                pair sexp string (section_sexp, Path.to_absolute_filename path)
-              | _ -> assert false)
-          in
-          Dune_lang.List (sexp_sections :: sections)
-        | x -> x)
-  ;;
-
-  let process_dune_package ~get_location ic =
+  let process_dune_package ~get_location ic ~src =
     let lb = Lexing.from_channel ic in
-    let dp =
-      Dune_lang.Parser.parse ~mode:Many lb |> List.map ~f:Dune_lang.Ast.remove_locs
-    in
-    (* replace sites with external path in the file *)
-    let dp = replace_sites ~get_location dp in
-    (* replace version if needed in the file *)
-    let need_version =
-      not
-        (List.exists dp ~f:(function
-          | Dune_lang.List (Atom (A "version") :: _)
-          | Dune_lang.List [ Atom (A "use_meta"); Atom (A "true") ]
-          | Dune_lang.List [ Atom (A "use_meta") ] -> true
-          | _ -> false))
-    in
-    let callback ?version ppf =
-      let dp =
-        match version with
-        | Some version ->
-          let version =
-            Dune_lang.List
-              [ Dune_lang.atom "version"; Dune_lang.atom_or_quoted_string version ]
-          in
-          (match dp with
-           | lang :: name :: rest -> lang :: name :: version :: rest
-           | [ lang ] -> [ lang; version ]
-           | [] -> [ version ])
-        | _ -> dp
+    let dune_version = Dune_lang.Syntax.greatest_supported_version_exn Stanza.syntax in
+    match Dune_package.Or_meta.parse src lb |> User_error.ok_exn with
+    | Use_meta ->
+      let callback ?version:_ ppf = Dune_package.Or_meta.pp_use_meta ~dune_version ppf in
+      Some { need_version = false; callback }
+    | Dune_package dp ->
+      (* replace sites with external path in the file *)
+      let dp, replace_info = Dune_package.replace_site_sections ~get_location dp in
+      (* replace version if needed in the file *)
+      let need_version = Option.is_none dp.version in
+      let callback ?version ppf =
```

Comment:
We should get rid of this optional argument. Not necessarily in this PR, but at least leave a note.